File: Chap5-MVCAdvanced/5-19CustomReadOnlyTreeModel/personmodel.cpp
#include "personmodel.h"
#include <QTextStream>
#include <QDebug>

PersonModel::PersonModel(QObject *parent) : QAbstractItemModel(parent)
{
    rootPerson = new Person("Names","Proffession");
    readFile();
}

PersonModel::~PersonModel()
{
    //Memory management
    delete rootPerson;
}

QModelIndex PersonModel::index(int row, int column, const QModelIndex &parent) const
{
    if (!hasIndex(row, column, parent))
        return QModelIndex();
    Person * parentPerson;

    if(!parent.isValid()){
        parentPerson = rootPerson;
    }else{
        parentPerson = static_cast<Person*>(parent.internalPointer());
    }

    Person * chilPerson = parentPerson->child(row);
    if(chilPerson)
        return createIndex(row,column,chilPerson);
    return QModelIndex();
}

QModelIndex PersonModel::parent(const QModelIndex &child) const
{
    if (!child.isValid())
        return QModelIndex();

    Person * childPerson = static_cast<Person*>(child.internalPointer());
    Person * parentPerson = childPerson->parentPerson();
    if(parentPerson== rootPerson)
        return QModelIndex();
    return createIndex(parentPerson->row(),0,parentPerson);
}

int PersonModel::rowCount(const QModelIndex &parent) const
{
    Person * parentPerson;
    /*
         * Only the first columns can have children
         * */
    if(parent.column() > 0)
        return 0;

    if(!parent.isValid()){
        parentPerson = rootPerson;
    }else{
        parentPerson = static_cast<Person*>(parent.internalPointer());
    }

    return  parentPerson->childCount();

}

int PersonModel::columnCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return  2;
}

QVariant PersonModel::data(const QModelIndex &index, int role) const
{
    if (!index.isValid())
        return QVariant();

    //No need to handle edit role here, the model is not editable anyway.
    if (role != Qt::DisplayRole)
        return QVariant();

    Person * person = static_cast<Person*>(index.internalPointer());

    return person->data(index.column());

}

QVariant PersonModel::headerData(int section, Qt::Orientation orientation, int role) const
{
    if(orientation == Qt::Horizontal && (role == Qt::DisplayRole)){
        return  rootPerson->data(section);
    }
    return QVariant();
}

Qt::ItemFlags PersonModel::flags(const QModelIndex &index) const
{
    if (!index.isValid())
        return nullptr;
    return QAbstractItemModel::flags(index);
}




void PersonModel::readFile()
{
    QString filename = ":/data/familytree1.txt";
    QFile inputFile(filename);
    if(inputFile.open(QIODevice::ReadOnly)){
        int lastIndentation = 0;
        Person * lastParent = rootPerson;
        Person * lastPerson = nullptr;

        QTextStream in(&inputFile);

        while (!in.atEnd()) {

            QString line = in.readLine();

            int currentIndentation = line.count("\t");

            qDebug() << "Read line : " << line << " tab count : " << QString::number(currentIndentation);

            QStringList infoList = getNamesAndProffession(line);


            int diffIndent = currentIndentation - lastIndentation;

            if(diffIndent == 0){
                //First Level Person
                Person * person = new Person(infoList[0],infoList[1],lastParent);
                lastParent->appendChild(person);
                lastPerson = person;

            }else if(diffIndent > 0){
                //Move the parent down
                lastParent = lastPerson;
                Person * person = new Person(infoList[0],infoList[1],lastParent);
                lastParent->appendChild(person);
                lastPerson = person;
            }else{
                //Move up the parent chain looking the  correct parent
                int iterations = - diffIndent;
                for(int i = 0; i < iterations ; i++){
                    lastParent = lastParent->parentPerson();
                }
                Person * person = new Person(infoList[0],infoList[1],lastParent);
                lastParent->appendChild(person);
                lastPerson = person;
            }
            lastIndentation = currentIndentation;
        }
        inputFile.close();
    }

}


QStringList PersonModel::getNamesAndProffession(QString txtString)
{
    //Get rid of the spaces
    QString cleanedUpStr = txtString.trimmed();
    QStringList split = cleanedUpStr.split("(");
    split[1].chop(1);
    return split;
}

File: Chap5-ModelViewArchitectureAdvanced/5-19CustomReadOnlyTreeModel/personmodel.py
import os
from PySide6.QtCore import QAbstractItemModel, QModelIndex, Qt, QByteArray
from person import Person
import resources_rc

class PersonModel(QAbstractItemModel):
    def __init__(self, parent=None):
        """
        Initialize the PersonModel with a root person
        
        :param parent: Parent QObject
        """
        super().__init__(parent)
        self.root_person = Person("Names", "Profession")
        self.filename = ":/data/familytree1.txt"
        self.read_file()
        
        # For Qt Quick TreeView, expose role names
        self._role_names = {
            Qt.DisplayRole: QByteArray(b'display'),
            Qt.UserRole: QByteArray(b'profession')
        }

    def roleNames(self):
        """
        Override roleNames to provide mapping for QML
        
        :return: Dictionary of role names
        """
        return self._role_names

    def index(self, row, column, parent=QModelIndex()):
        """
        Create a model index for a given row and column
        
        :param row: Row of the item
        :param column: Column of the item
        :param parent: Parent model index
        :return: Model index for the specified item
        """
        if not self.hasIndex(row, column, parent):
            return QModelIndex()

        parent_person = parent.internalPointer() if parent.isValid() else self.root_person
        child_person = parent_person.child(row)

        return self.createIndex(row, column, child_person) if child_person else QModelIndex()

    def parent(self, child):
        """
        Get the parent model index for a given child index
        
        :param child: Child model index
        :return: Parent model index
        """
        if not child.isValid():
            return QModelIndex()

        child_person = child.internalPointer()
        parent_person = child_person.parent_person()

        if parent_person == self.root_person:
            return QModelIndex()

        return self.createIndex(parent_person.row(), 0, parent_person)

    def rowCount(self, parent=QModelIndex()):
        """
        Get the number of rows for a given parent
        
        :param parent: Parent model index
        :return: Number of rows
        """
        if parent.column() > 0:
            return 0

        parent_person = parent.internalPointer() if parent.isValid() else self.root_person
        return parent_person.child_count()

    def columnCount(self, parent=QModelIndex()):
        """
        Get the number of columns
        
        :param parent: Parent model index
        :return: Number of columns
        """
        return 2

    def data(self, index, role=Qt.DisplayRole):
        """
        Get data for a given index and role
        
        :param index: Model index
        :param role: Display role
        :return: Data for the specified index and role
        """
        if not index.isValid():
            return None

        person = index.internalPointer()
        
        if role == Qt.DisplayRole and index.column() == 0:
            return person.data(0)
        elif role == Qt.UserRole or (role == Qt.DisplayRole and index.column() == 1):
            return person.data(1)
            
        return None

    def headerData(self, section, orientation, role=Qt.DisplayRole):
        """
        Get header data
        
        :param section: Section number
        :param orientation: Horizontal or vertical orientation
        :param role: Display role
        :return: Header data
        """
        if orientation == Qt.Horizontal and role == Qt.DisplayRole:
            return "Name" if section == 0 else "Profession"
        return None

    def flags(self, index):
        """
        Get item flags for a given index
        
        :param index: Model index
        :return: Item flags
        """
        if not index.isValid():
            return Qt.NoItemFlags
        return super().flags(index)

    def read_file(self):
        """
        Read and parse the family tree file
        Parse the indentation-based hierarchical text file
        """
        last_indentation = 0
        last_parent = self.root_person
        last_person = None

        # Use QFile for resource file
        from PySide6.QtCore import QFile, QIODevice, QTextStream
        file = QFile(self.filename)
        
        if file.open(QIODevice.ReadOnly | QIODevice.Text):
            text_stream = QTextStream(file)
            
            while not text_stream.atEnd():
                line = text_stream.readLine()
                
                current_indentation = line.count('\t')
                names, profession = self.parse_line(line.strip())

                diff_indent = current_indentation - last_indentation

                if diff_indent == 0:
                    # Sibling level
                    person = Person(names, profession, last_parent)
                    last_parent.append_child(person)
                    last_person = person
                elif diff_indent > 0:
                    # Child level
                    last_parent = last_person
                    person = Person(names, profession, last_parent)
                    last_parent.append_child(person)
                    last_person = person
                else:
                    # Move up the parent chain
                    iterations = -diff_indent
                    for _ in range(iterations):
                        last_parent = last_parent.parent_person()
                    
                    person = Person(names, profession, last_parent)
                    last_parent.append_child(person)
                    last_person = person

                last_indentation = current_indentation

            file.close()

    def parse_line(self, line):
        """
        Parse a line into names and profession
        
        :param line: Line from the input file
        :return: Tuple of (names, profession)
        """
        parts = line.split('(')
        names = parts[0].strip()
        profession = parts[1].rstrip(')').strip()
        return names, profession